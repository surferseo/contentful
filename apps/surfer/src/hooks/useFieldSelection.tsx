import { EntryFieldAPI } from '@contentful/app-sdk';
import { useCallback, useMemo, useState } from 'react';
import { FieldSelection } from '../components/FieldSelection';

export const useFieldSelection = (entryFields: EntryFieldAPI[]) => {
  const richTextFields = useMemo(() => Object.values(entryFields).filter((field) => field.type === 'RichText'), [entryFields]);
  const [selectedFields, setSelectedFields] = useState<EntryFieldAPI[]>(richTextFields);

  const isSelected = useCallback((field: EntryFieldAPI) => selectedFields.some((selectedField) => selectedField.id === field.id), [selectedFields]);

  const toggleSelection = useCallback(
    (field?: EntryFieldAPI) => {
      if (field && isSelected(field)) {
        return setSelectedFields((selectedFields) => selectedFields.filter((selectedField) => selectedField.id !== field.id));
      } else if (field) {
        setSelectedFields((selectedFields) => [...selectedFields, field]);
      }
    },
    [isSelected]
  );

  const SelectionComponent =
    richTextFields.length > 1
      ? () => <FieldSelection selectedFields={selectedFields} richTextFields={richTextFields} toggleSelection={toggleSelection} isSelected={isSelected} />
      : () => null;

  return [selectedFields, SelectionComponent, richTextFields] as const;
};
