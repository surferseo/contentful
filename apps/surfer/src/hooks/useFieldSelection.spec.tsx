import { render, renderHook } from '@testing-library/react';
import { useFieldSelection } from './useFieldSelection';
import { EntryFieldAPI } from '@contentful/app-sdk';

describe('useFieldSelection', () => {
  const fakeFields: EntryFieldAPI[] = [
    { type: 'RichText', id: 'test-id' },
    { type: 'RichText', id: 'test-id2' },
    { type: 'Text', id: 'test-id3' },
    { type: 'Integer', id: 'test-id3' },
  ] as any;

  describe('selectedFields', () => {
    it('by default is an empty array if there are no rich text fields', async () => {
      const { result } = renderHook(() => useFieldSelection(fakeFields.slice(2)));

      expect(result.current[0]).toEqual([]);
    });

    it('by default is an array of all rich text fields if there are multiple', async () => {
      const { result } = renderHook(() => useFieldSelection(fakeFields));

      expect(result.current[0]).toEqual(fakeFields.slice(0, 2));
    });
  });

  describe('SelectionComponent', () => {
    it('is an empty placeholder if there are no rich text componets', async () => {
      const { result } = renderHook(() => useFieldSelection([]));

      const SelectionComponent = result.current[1];

      expect(render(<SelectionComponent />).container).toBeEmptyDOMElement();
    });

    it("is an empty placeholder if there's only one rich text field", async () => {
      const { result } = renderHook(() => useFieldSelection(fakeFields.slice(0, 1)));

      const SelectionComponent = result.current[1];

      expect(SelectionComponent()).toBe(null);
    });

    it('renders a FieldSelection component if there are multiple rich text fields', async () => {
      const { result } = renderHook(() => useFieldSelection(fakeFields));

      const SelectionComponent = result.current[1];

      expect(SelectionComponent()).toMatchSnapshot();
    });
  });

  describe('richTextFields', () => {
    it('is an empty array if there are no rich text fields', async () => {
      const { result } = renderHook(() => useFieldSelection(fakeFields.slice(2)));

      expect(result.current[2]).toEqual([]);
    });

    it('is an array of all rich text fields if there are multiple', async () => {
      const { result } = renderHook(() => useFieldSelection(fakeFields));

      expect(result.current[2]).toEqual(fakeFields.slice(0, 2));
    });
  });
});
