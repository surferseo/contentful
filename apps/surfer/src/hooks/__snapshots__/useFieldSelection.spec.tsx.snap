// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFieldSelection SelectionComponent renders a FieldSelection component if there are multiple rich text fields 1`] = `
<ForwardRef
  isSelected={[Function]}
  richTextFields={
    Array [
      Object {
        "id": "test-id",
        "type": "RichText",
      },
      Object {
        "id": "test-id2",
        "type": "RichText",
      },
    ]
  }
  selectedFields={
    Array [
      Object {
        "id": "test-id",
        "type": "RichText",
      },
      Object {
        "id": "test-id2",
        "type": "RichText",
      },
    ]
  }
  toggleSelection={[Function]}
/>
`;
